runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/StreamingEngineClient.java

```diff
@@ -0,0 +1,444 @@
+      GrpcWindmillStreamFactory streamFactory,
+      WorkItemProcessor processWorkItem,
+      WindmillStubFactory stubFactory,
+      GetWorkBudgetDistributor getWorkBudgetDistributor,
+      GrpcDispatcherClient dispatcherClient,
+      long clientId) {
+    StreamingEngineClient streamingEngineClient =
+        new StreamingEngineClient(
+            jobHeader,
+            totalGetWorkBudget,
+            connections,
+            streamFactory,
+            processWorkItem,
+            stubFactory,
+            getWorkBudgetDistributor,
+            dispatcherClient,
+            clientId);
+    streamingEngineClient.startGetWorkerMetadataStream();
+    streamingEngineClient.startWorkerMetadataConsumer();
+    return streamingEngineClient;
+  }
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void startWorkerMetadataConsumer() {
+    newWorkerMetadataConsumer.submit(
+        () -> {
+          while (true) {
+            Optional.ofNullable(newWindmillEndpoints.poll())
+                .ifPresent(this::consumeWindmillWorkerEndpoints);
+          }
+        });
+  }
+
+  /**
+   * Waits for the initial worker metadata to be consumed. Does nothing if this has already been
+   * called.
+   *
+   * @throws IllegalArgumentException if trying to start before {@link #connections} are set with
+   *     {@link GetWorkerMetadataStream}.
+   */
+  public void waitForFirstWorkerMetadata() {
+    // Do nothing if we have already initialized the initial streams.
+    if (started.compareAndSet(false, true)) {
+      return;
+    }
+
+    // This will block if we have not received the first response from GetWorkerMetadata.
+    try {
+      getWorkerMetadataReady.await();
+    } catch (InterruptedException e) {
+      throw new StreamingEngineClientException(
+          "Error occurred waiting for StreamingEngine backend endpoints.", e);
+    }
+  }
+
+  @VisibleForTesting
+  void finish() {
+    if (!started.compareAndSet(true, false)) {
+      return;
+    }
+    getWorkerMetadataStream.get().close();
+    getWorkBudgetRefresher.stop();
+    newWorkerMetadataPublisher.shutdownNow();
+    newWorkerMetadataConsumer.shutdownNow();
+  }
+
+  private void onFirstWorkerMetadata() {
+    StreamingEngineConnectionState currentConnectionsState = connections.get();
+    Preconditions.checkState(
+        !StreamingEngineConnectionState.EMPTY.equals(currentConnectionsState),
+        "Cannot start streams without connections.");
+
+    LOG.info("Starting initial GetWorkStreams with connections={}", currentConnectionsState);
+    ImmutableCollection<WindmillStreamSender> windmillStreamSenders =
+        currentConnectionsState.windmillStreams().values();
+    getWorkBudgetDistributor.distributeBudget(
+        currentConnectionsState.windmillStreams().values(), totalGetWorkBudget);
+    lastBudgetRefresh.compareAndSet(Instant.EPOCH, Instant.now());
+    windmillStreamSenders.forEach(WindmillStreamSender::startStreams);
```

Comment:
where do streams get started on non-first metadata?

Reply:
added starting the stream when it is created in `createAndStartWindmillStreamSenderFor(
      WindmillConnection connection)`.